make listening socket creation optional for userspace path manager plugins

marco-a-itl opened the pull request:
Proposal to resolve #296

matttbe commented on include/mptcpd/path_manager.h:
```diff
@@ -118,13 +118,15 @@ MPTCPD_API bool mptcpd_pm_ready(struct mptcpd_pm const *pm);
                                   mptcpd_aid_t id,
-                                  mptcpd_token_t token);
+                                  mptcpd_token_t token,
+                                  bool nolst);
```
This will break the API, no?

Maybe it would be safer to export a new function? `mptcpd_pm_add_addr_no_listener()`?

Internally, they will share the same code, but we don't break the API, not to force people to check the version to see what parameter to pass to a function: they can use `mptcpd_pm_add_addr_no_listener()` if it is available, and if they want to, that seems easier I think.

matttbe reviewed (changes requested):
Thank you for the PR! It looks good to me, but I think it will break the API, and it looks like we can avoid that easily, no?

marco-a-itl commented:
Modified the PR so that a new `mptcpd_pm_add_addr_no_listener` function is defined.
Fixed also `mptcp_org_add_addr` compile error for out-of-tree kernel.

matttbe commented on src/netlink_pm_upstream.c:
```diff
@@ -219,20 +219,21 @@ static int send_add_addr(struct mptcpd_pm *pm,
-         * @todo This should be optional.
-         */
-        int const r = mptcpd_lm_listen(pm->lm, addr);
+        if (!nolst) {
```
nit: now that _no listener_ parameter is only internal, it might be clearer to use the "positive" form, and avoid a "double negation" here: _if not no listener_

    if (listener) {
         (...)

matttbe commented on lib/path_manager.c:
```diff
@@ -43,6 +43,31 @@ static bool is_pm_ready(struct mptcpd_pm const *pm, char const *fname)
         return ready;
 }
 
+static int do_pm_add_addr(struct mptcpd_pm *pm,
```
nit: because it is specific to `mptcpd_pm_add_addr*()` and not used elsewhere, should this be kept just above them?

matttbe reviewed (changes requested):
Thank you! It looks good to me, just two last small requests if that's OK for you.

@mjmartineau / @ossama-othman : because this modifies the public API, do you mind having a quick look please?

marco-a-itl commented:
Modified the PR to avoid double negation check.
`do_pm_add_addr` is just above `mptcpd_pm_add_addr*()`.

matttbe reviewed (approved):
Thank you for the update! It looks good to me, but I prefer to wait for Ossama or Mat's feedback (if they are available).

matttbe commented:
It looks like the issue with the [CI](https://github.com/multipath-tcp/mptcpd/actions/runs/10286346028/job/28467051717?pr=297) is due to an [API breakage](https://git.kernel.org/pub/scm/libs/ell/ell.git/commit/?id=883bc6a90b4a9c654fdf2540f02447a7ca567e04) on ELL side. I just opened a new bug report as it is not due to these modifications, anybody can fix it: #299

matttbe commented:
@ossama-othman: if you have the opportunity, do you mind having a quick look because this PR modifies the public API please? :)

ossama-othman commented:
@matttbe Sure!  I'll take a look today.

@marco-a-itl Thank you very much for the PR!

ossama-othman reviewed (approved):
It looks good to me. Thank you very much!

Perhaps not in this PR, but when releasing a new version of mptcpd the `NEWS` file could be updated with a description of the API addition, as well bumping the libtool versioning accordingly.  From `configure.ac`:

```sh
# ---------------------------------------------------------------
# Libtool Based Library Versioning
# ---------------------------------------------------------------
# Source     changed:  REVISION++
# Interfaces changed:  CURRENT++ REVISION=0
#            added:    CURRENT++ REVISION=0 AGE++
#            removed:  CURRENT++ REVISION=0 AGE=0
LIB_CURRENT=3
LIB_REVISION=0
LIB_AGE=0
```

matttbe commented:
> It looks good to me. Thank you very much!

Thank you! Just squashed and merged!

> Perhaps not in this PR, but when releasing a new version of mptcpd the `NEWS` file could be updated with a description of the API addition, as well bumping the libtool versioning accordingly.

OK, noted!

> From `configure.ac`:
> 
> ```shell
> # ---------------------------------------------------------------
> # Libtool Based Library Versioning
> # ---------------------------------------------------------------
> # Source     changed:  REVISION++
> # Interfaces changed:  CURRENT++ REVISION=0
> #            added:    CURRENT++ REVISION=0 AGE++
> #            removed:  CURRENT++ REVISION=0 AGE=0
> LIB_CURRENT=3
> LIB_REVISION=0
> LIB_AGE=0
> ```

So here, because only new things have been added, I would need to update both `CURRENT` and `AGE`, to say that this lib supports everything implemented from `CURRENT-AGE`?
